File: unity/oelMVCS/Assets/XTC/oelMVCS/Scripts/Service/ServiceCenter.cs
using System.Collections.Generic;

namespace XTC.oelMVCS
{
    public class ServiceCenter 
    {
        public ServiceCenter(Board _board)
        {
            board_ = _board;
        }

        public Error Register(string _uuid, Service.Inner _inner)
        {
            board_.getLogger().Info("register {0}", _uuid);
            if (units_.ContainsKey(_uuid))
                return Error.NewAccessErr("{0} exists", _uuid);
            units_[_uuid] = _inner;
            return Error.OK;
        }

        public Error Cancel(string _uuid)
        {
            board_.getLogger().Info("cancel {0}", _uuid);
            if (!units_.ContainsKey(_uuid))
                return Error.NewAccessErr("{0} not found", _uuid);
            units_.Remove(_uuid);
            return Error.OK;
        }

        public Service.Inner FindUnit(string _uuid)
        {
            Service.Inner inner = null;
            units_.TryGetValue(_uuid, out inner);
            return inner;
        }

        public void Setup()
        {
            foreach (Service.Inner inner in units_.Values)
            {
                inner.Setup(board_);
            }
        }

        public void Dismantle()
        {
            foreach (Service.Inner inner in units_.Values)
            {
                inner.Dismantle();
            }
        }

        private Dictionary<string, Service.Inner> units_ = new Dictionary<string, Service.Inner>();
        private Board board_ = null;
    }
}//namespace

File: python/oelMVCS/ServiceCenter.py
# -*- coding: utf-8 -*-

class ServiceCenter:

    def __init__(self, _framework):
         self.__framework = _framework
         self.services= {}

    def Register(self, _uuid, _service):
        self.__framework.logger.Info('register service %s'%(_uuid))
        if _uuid in self.services:
            raise KeyError('service is exists')
        self.services[_uuid] = _service

    def Cancel(self, _uuid):
        self.__framework.logger.Info('cancel service %s'%(_uuid))
        if not _uuid in self.services:
            raise KeyError('service not found')
        del self.services[_uuid] 

    def Find(self, _uuid):
        return self.services[_uuid]

    def Setup(self):
        for v in self.services.values():
            v.Setup(self.__framework)

    def Dismantle(self):
        for v in self.services.values():
            v.Dismantle()

